Change hdc1008 driver into ti_hdc and add support for hdc1080

nikooiko opened the pull request:
Hello everyone, in this PR I added support for the HDC1080 Temp/Rh sensor.

I'm currently using this driver locally and I needed the driver to also support raw measurements. In order to do that, I added two new channels for raw Temp/RH. Is there a better way to implement this?

This PR adds on top of the https://github.com/zephyrproject-rtos/zephyr/pull/11265.

Signed-off-by: Nikos Oikonomou <nikoikonomou92@gmail.com>

nikooiko commented:
This PR must not be merged until https://github.com/zephyrproject-rtos/zephyr/pull/11265 is merged.

erwango commented on samples/boards/nucleo_l496zg/sensors/hdc1080/nucleo_l496zg.overlay:
```diff
@@ -0,0 +1,15 @@
+&i2c1 {
+
+	hdc1080: hdc1080@40 {
+		compatible = "ti,hdc1080";
```
Please add matching yaml binding under dts/bindings/sensor

nikooiko replied:
Done!

erwango commented on samples/boards/nucleo_l496zg/sensors/hdc1080/dts_fixup.h:
```diff
@@ -0,0 +1,15 @@
+/*
```
Remove this file

nikooiko replied:
Done!

erwango commented on samples/boards/nucleo_l496zg/sensors/hdc1080/src/main.c:
```diff
@@ -0,0 +1,46 @@
+void main(void)
+{
+	printk("Running on %s!\n", CONFIG_ARCH);
+	struct device *dev = device_get_binding(DT_HDC1080_NAME);
```
Replace DT_HDC1080_NAME with DT_TI_HDCXXXX_0_LABEL

nikooiko replied:
Done!

erwango commented on samples/boards/nucleo_l496zg/sensors/README.rst:
```diff
@@ -0,0 +1,53 @@
+
+To use this sample, the following hardware is required:
+
+* Nucleo STM32L496ZG
```
Is it? 
Won't it work with any board supporting zephyr with an I2C port?

nikooiko replied:
You are right, I reworked the README and now just mention this nucleo as a tested board.

erwango reviewed (changes requested):
Thanks for the change.
Though, since this sensor is not available by default on nucleo_l496zg, so please introduce the sample under sample/sensor/ and mention nucleo_l496zg as an exmaple rather than the only board supporting this sensor

nikooiko commented on samples/basic/blink_led/README.rst:
```diff
@@ -25,6 +25,11 @@ Nucleo_F103RB
 
+Nucleo_L496ZG
+=============
+No special board setup is necessary because there are three on-board leds (red,
```
This will be resolved directly at the source PR (https://github.com/zephyrproject-rtos/zephyr/pull/11265)

nikooiko replied:
Done!

nikooiko commented on samples/boards/nucleo_l496zg/sensors/README.rst:
```diff
@@ -0,0 +1,53 @@
+.. _sensor_samples:
```
Instead it's renamed to hdc1080-sample

nikooiko commented on samples/boards/nucleo_l496zg/sensors/README.rst:
```diff
@@ -0,0 +1,53 @@
+
+.. code-block:: bash
+
+    $ZEPHYR_BASE/scripts/sanitycheck -T $ZEPHYR_BASE/samples/boards/nucleo_l496zg/sensors/hdc1080/ -p nucleo_l496zg --device-testing --device-serial /dev/ttyUSB0
```
Should I rephrase it somehow or do something else?

jfischer-no reviewed (changes requested):
Please use HDC1008 driver `drivers/sensors/hdc1008`, it has the same register setting and similar functionality. There is no reason for code duplication.

nikooiko commented:
Most sensor drivers from the same company are similar and they differ in some details (e.g. Bosch, TI, Alphasense). I don't think that this should stop us from integrating multiple sensors of the same type. If you prefer I could try to merge both drivers into one ( HDC10xx), but this would result in more complex drivers, with mixed ifdefs and not so clear flow. For example, although they are similar, HDC1008 has the capability to notify the host MCU with its RDY pin, while HDC1080 doesn't.

nikooiko commented:
Personally, I would prefer to keep it clean by having two discrete drivers, but if you decide that this is demanding for the merge request to finish, I'll continue with the merge.

But if we decide to merge them, we should change the name to something like HDC10xx because otherwise, it will be confusing when someone is trying to understand whether he can use this driver or not for sensors other than the HDC1008.

jfischer-phytec-iot commented:
>Personally, I would prefer to keep it clean by having two discrete drivers, but if you decide that this is demanding for the merge request to finish, I'll continue with the merge.

Just add option "Wait for the measurements to complete, based on the conversion time" to hdc1008 driver, revise hardcoded Device ID and it will work with your device. Again, what for reason would be there to have identically driver :question:

gon1332 commented:
> > Personally, I would prefer to keep it clean by having two discrete drivers, but if you decide that this is demanding for the merge request to finish, I'll continue with the merge.
> 
> Just add option "Wait for the measurements to complete, based on the conversion time" to hdc1008 driver, revise hardcoded Device ID and it will work with your device. Again, what for reason would be there to have identically driver question

@jfischer-phytec-iot It would be a good decision to have this option, just not to have duplicate code.
However, I also believe that the naming should be different. Since the _DTS_ includes the _hardware features_ of a system, the names on them should reflect the exact hardware components soldered.
Something like HDC10XX would be fine.

nikooiko commented:
On Monday, I'll merge the hdc1080 and hdc1008 drivers into one hdc10xx driver as described above.

Renaming merge request to add hdc10xx sensor (and removing WIP because https://github.com/zephyrproject-rtos/zephyr/pull/11265 is merged).

jfischer-phytec-iot commented:
>Again, I agree with you, as you explained hdc10xx would rule out hdc20xx, but again the current name is specific for the hdc1008 version. Wouldn't be more appropriate to rename it to something even more generic, e.g. ti-hdc or smth?

`ti_hdc` would indeed fit better. @galak Your opinion regarding renaming?

nikooiko commented:
I believe requested changes are finished.

Sorry for the subsequent force pushes... :)

@jfischer-phytec-iot are you ok with the final version?

jfischer-phytec-iot commented on samples/boards/arduino_101/environmental_sensing/README.rst:
```diff
@@ -25,7 +25,7 @@ To use this sample, the following hardware is required:
 * `Grove LCD module`_
 * `Grove Base Shield`_ [Optional]
-* `HDC1008 Sensor`_
+* `TI_HDC Sensor`_
```
It is still HDC1008 sensor. And what has it to do with _samples: added ti_hdc_ ?

nikooiko replied:
Noted

jfischer-phytec-iot commented on samples/boards/reel_board/mesh_badge/src/periphs.c:
```diff
@@ -24,15 +24,15 @@ struct led_device_info {
-	{ NULL, LED1_GPIO_CONTROLLER, LED1_GPIO_PIN }, /* red front LED */
-	{ NULL, LED2_GPIO_CONTROLLER, LED2_GPIO_PIN }, /* green front LED */
-	{ NULL, LED3_GPIO_CONTROLLER, LED3_GPIO_PIN }, /* blue front LED */
+	{ NULL, LED0_GPIO_CONTROLLER, LED0_GPIO_PIN },  /* green back LED */
```
?

nikooiko replied:
Changed by uncrustify, will revert back.

jfischer-phytec-iot commented on samples/sensor/ti_hdc/README.rst:
```diff
@@ -0,0 +1,81 @@
+Requirements
+************
+
+This sample needs an TI_HDC sensor connector to the target
```
There is no TI_HDC sensor, "sample needs a compatible sensor like HDC1010, HDC1008"

nikooiko replied:
Noted

jfischer-no reviewed (changes requested):
>@jfischer-phytec-iot are you ok with the final version?

no.

Please split your commits in several parts, like renaming, add _"Wait for the measurements to complete, based on the conversion time"_ functionality... Remove any changes that do not belong to this PR.

nikooiko commented on drivers/sensor/ti_hdc/ti_hdc.c:
```diff
@@ -0,0 +1,209 @@
+		val->val1 = (s32_t)(tmp >> 16) - 40;
+		val->val2 = (1000000 * (tmp & 0xFFFF)) >> 16;
+		break;
+	case SENSOR_CHAN_AMBIENT_TEMP_RAW:
```
Ok I will extract this functionality and after this merge request is merged I will create a new one just for the raw sensor values retrieval.

nikooiko commented on drivers/sensor/ti_hdc/ti_hdc.c:
```diff
@@ -0,0 +1,209 @@
+#endif  /* CONFIG_TI_HDC_DRDY */
+
+	/* configure the sensor */
+	helper = HDC_RH_RES_14BIT | HDC_TEMP_RES_14BIT |
```
You're right these are the default values, but this also acts as a demonstration about how someone can change the configuration of the sensor. We can either keep it to this merge request, or I could create a new one after this is merged, to add this only. Whatever suits you best.

jfischer-phytec-iot replied:
It does not demonstrate that one can change the configuration. The sensors are almost perfect, after the power on they are just ready and low power per design. I see no need to write the default settings again. It costs time and energy every time the sensor is turned on.

nikooiko replied:
The TI offers these configuration options in case someone needs to operate them in a different way. We might not need it now but in the future, someone might need to change it to operate the sensor in a different way.

If the cost is the problem, then I disagree that it affects the power efficiency of any device out there that uses this sensor. It's just a one-off I2C write operation.

jfischer-phytec-iot replied:
This commit is not necessary, please drop it. It still not demonstrate that one can change the configuration, It just sets default POR values.

nikooiko commented on drivers/sensor/ti_hdc/ti_hdc.h:
```diff
@@ -0,0 +1,45 @@
+
+#include <kernel.h>
+
+#define TI_HDC_REG_TEMP 0x0
```
This isn't enforced by the uncrustify tool? I don't know much about uncrustify and how it applies the coding style.

nikooiko commented on include/sensor.h:
```diff
@@ -79,6 +79,8 @@ enum sensor_channel {
 	/** Ambient temperature in degrees Celsius. */
 	SENSOR_CHAN_AMBIENT_TEMP,
+	/** Ambient temperature in raw format. */
+	SENSOR_CHAN_AMBIENT_TEMP_RAW,
```
Resolved at previous discussion. Thanks

nikooiko commented on dts/bindings/sensor/ti,hdc.yaml:
```diff
@@ -4,21 +4,21 @@
 #
 ---
-title: HDC1008 Temperature and Humidity Sensor
+title: TI_HDC Temperature and Humidity Sensor
```
Noted

nikooiko commented on drivers/sensor/ti_hdc/ti_hdc.c:
```diff
@@ -0,0 +1,209 @@
+	k_sem_take(&drv_data->data_sem, K_FOREVER);
+#else
+	/* wait for the conversion to finish */
+	k_sleep(13); /* For 14bit conversion RH needs 6.5ms and Temp 6.35ms */
```
Noted

nikooiko commented on drivers/sensor/ti_hdc/Kconfig:
```diff
@@ -0,0 +1,17 @@
+	help
+	  Enable driver for TI_HDC temperature and humidity sensors.
+
+config TI_HDC_DRDY
```
Noted

nikooiko commented on boards/arm/reel_board/reel_board.dts:
```diff
@@ -95,10 +95,10 @@
+	ti_hdc@43 {
+		compatible = "ti,hdc";
 		reg = <0x43>;
-		label = "HDC1010";
```
Noted

nikooiko commented on boards/arm/reel_board/reel_board.dts:
```diff
@@ -95,10 +95,10 @@
 	};
 
-	hdc1008@43 {
-		compatible = "ti,hdc1008","ti,hdc1010";
```
Noted

nikooiko committed:
sensors: renamed hdc1008 driver into ti_hdc

Hdc1008 driver is renamed into ti_hdc to prepare it to support all
available Texas Instruments HDC sensors (e.g. hdc1080, hdc2080).

Signed-off-by: Nikos Oikonomou <nikoikonomou92@gmail.com>

nikooiko committed:
samples: added sample for ti_hdc

Added a new sample for TI_HDC driver, tested with nucleo and hdc1080.

Signed-off-by: Nikos Oikonomou <nikoikonomou92@gmail.com>

jfischer-phytec-iot commented on samples/sensor/ti_hdc/README.rst:
```diff
@@ -0,0 +1,81 @@
+Description
+***********
+
+This sample application periodically takes Temp/RH measurements
```
Temperature and Humidity

nikooiko replied:
Noted

jfischer-phytec-iot replied:
Not addressed

jfischer-phytec-iot commented on samples/sensor/ti_hdc/prj.conf:
```diff
@@ -0,0 +1,4 @@
+CONFIG_ASSERT=y # validation
+CONFIG_I2C=y
+CONFIG_SENSOR=y
+CONFIG_TI_HDC=y
```
The sample is unnecessary limited, with `CONFIG_GPIO=y` the sample can also be used for HDC1010 or HDC1008 sensors.

nikooiko replied:
Noted

jfischer-phytec-iot replied:
Not resolved!

jfischer-phytec-iot commented on samples/sensor/ti_hdc/src/main.c:
```diff
@@ -0,0 +1,41 @@
+		/* print the result */
+		printk("Temp = %d.%06d C, RH = %d.%06d %%\n",
+		       temp.val1, temp.val2, humidity.val1, humidity.val2);
+		k_sleep(K_SECONDS(1));
```
redundant?

nikooiko replied:
This sleep is just for not spamming stdout

jfischer-phytec-iot replied:
But below is the sleep for 10 seconds.

nikooiko replied:
Ah sorry didn't notice that it was a duplicate. Will be deleted

jfischer-phytec-iot replied:
Not addressed

jfischer-no reviewed (changes requested):
Thanks, looks better, a few more comments

nikooiko commented on drivers/sensor/ti_hdc/ti_hdc.c:
```diff
@@ -37,23 +39,35 @@ static int ti_hdc_sample_fetch(struct device *dev, enum sensor_channel chan)
+#if defined(DT_TI_HDC_0_DRDY_GPIOS_CONTROLLER)
 	k_sem_take(&drv_data->data_sem, K_FOREVER);
+#else
+	/* wait for the conversion to finish */
```
Will be deleted

nikooiko commented on drivers/sensor/ti_hdc/ti_hdc.c:
```diff
@@ -37,23 +39,35 @@ static int ti_hdc_sample_fetch(struct device *dev, enum sensor_channel chan)
+	k_sleep(HDC_CONVERSION_TIME);
+#endif
 
+	/* read the measurements */
```
Will be deleted

nikooiko commented on drivers/sensor/ti_hdc/ti_hdc.h:
```diff
@@ -9,23 +9,30 @@
 #include <kernel.h>
 
-#define TI_HDC_I2C_ADDRESS	0x40
+#define TI_HDC_REG_TEMP         0x0
```
It's 1 tab or more? Sorry, my editor is usually set to use spaces.

nikooiko replied:
Uncrustify seems to affect spaces/tabs.

nikooiko replied:
Uncrustify seems to affect spaces/tabs.

nikooiko commented on drivers/sensor/ti_hdc/ti_hdc.c:
```diff
@@ -118,25 +133,27 @@ static int ti_hdc_init(struct device *dev)
+			DT_TI_HDC_0_BASE_ADDRESS,
+			TI_HDC_DEVID_REG);
+	if (helper != TI_HDC1000_DEVID && helper != TI_HDC1050_DEVID) {
 		LOG_ERR("Failed to get correct device ID");
```
Noted

nikooiko commented on drivers/sensor/ti_hdc/ti_hdc.c:
```diff
@@ -37,23 +39,35 @@ static int ti_hdc_sample_fetch(struct device *dev, enum sensor_channel chan)
-	gpio_pin_enable_callback(drv_data->gpio,
-				 DT_TI_HDC_0_DRDY_GPIOS_PIN);
+	/* clear sensor values */
+	drv_data->t_sample = 0;
```
Will be deleted

nikooiko commented on drivers/sensor/ti_hdc/ti_hdc.c:
```diff
@@ -108,6 +122,7 @@ static u16_t read16(struct device *dev, u8_t a, u8_t d)
 
 static int ti_hdc_init(struct device *dev)
 {
+	u16_t helper;
```
Noted

nikooiko commented on drivers/sensor/ti_hdc/ti_hdc.c:
```diff
@@ -37,23 +39,35 @@ static int ti_hdc_sample_fetch(struct device *dev, enum sensor_channel chan)
 		return -EIO;
 	}
 
+	/* store measurements to the driver */
```
Will be deleted

dbkinder reviewed (approved):
+1 for docs, thanks.

nikooiko commented:
@jfischer-phytec-iot can you please check my replies to continue with the changes?

gon1332 commented:
@jfischer-phytec-iot Any news about this PR?

jfischer-phytec-iot commented:
@gon1332 I did not forget it, will review and test asap

jfischer-phytec-iot commented on drivers/sensor/ti_hdc/ti_hdc.h:
```diff
@@ -9,23 +9,30 @@
-#define TI_HDC_REG_HUMIDITY	0x1
-#define TI_HDC_REG_MANUFID	0xFE
-#define TI_HDC_REG_DEVICEID	0xFF
+#define TI_HDC_MANUFID_REG      0xFE
```
The schema was TI_HDC_REG_*, why you changed it for TI_HDC_REG_DEVICEID?
This also leads to unnecessary changes [here](https://github.com/zephyrproject-rtos/zephyr/pull/12616/commits/0f7e9a48482fecc6cd41b24cd7c375919e447550#diff-7e009466fba8f8e3e9e60aeb0c4f2856L121)

jfischer-phytec-iot commented:
Please stop marking unresolved comments resolved.

nikooiko commented:
Noted. Can you answer my questions so I can proceed with the changes?

nikooiko commented:
The spaces/tabs are changed by the checkpatch, so I can't do something about it.

@jfischer-phytec-iot is everything else ok?

jfischer-phytec-iot commented:
>The spaces/tabs are changed by the checkpatch, so I can't do something about it.

Please fix it in editor.

@jfischer-phytec-iot is everything else ok?

If I did not miss anything, yes. Thanks for your contribution, it looks generally very good.

nikooiko commented:
Isn't it against the contribution guideline?

jfischer-phytec-iot commented on drivers/sensor/ti_hdc/ti_hdc.c:
```diff
@@ -18,8 +18,9 @@
 #define LOG_LEVEL CONFIG_SENSOR_LOG_LEVEL
```
You may wish to add your copyright notice as you extended the functionality of the driver.

jfischer-phytec-iot commented:
>Isn't it against the contribution guideline?

To use the editor? :smile:

nikooiko commented:
No, to bypass the checkpatch. If not then I guess that would work :P

nikooiko commented:
Added one tab after each definition (same as initial hdc1008.h). @jfischer-phytec-iot is it ok now?

nikooiko commented:
@jfischer-phytec-iot did you find some time to check it?

MaureenHelm commented on drivers/sensor/Kconfig:
```diff
@@ -59,7 +59,7 @@ source "drivers/sensor/fxos8700/Kconfig"
 source "drivers/sensor/grove/Kconfig"
 
-source "drivers/sensor/hdc1008/Kconfig"
+source "drivers/sensor/ti_hdc/Kconfig"
```
Please explain in your commit message why you are renaming this driver

MaureenHelm commented on drivers/sensor/ti_hdc/Kconfig:
```diff
@@ -8,6 +8,6 @@
 menuconfig TI_HDC
 	bool "Texas Instruments Temperature and Humidity Sensor (e.g. HDC1008)"
-	depends on I2C && HAS_DTS_I2C && GPIO && HAS_DTS_GPIO
+	depends on I2C && HAS_DTS_I2C
```
Why are you removing this dependency?

nikooiko replied:
Because it depends on the variation of the ti_hdc sensor and the GPIO interrupt functionality is optional. Of course, if you have a better idea I can change it.

MaureenHelm commented on drivers/sensor/ti_hdc/ti_hdc.c:
```diff
@@ -18,8 +18,9 @@
+#if defined(DT_TI_HDC_0_DRDY_GPIOS_CONTROLLER)
 static void ti_hdc_gpio_callback(struct device *dev,
-				  struct gpio_callback *cb, u32_t pins)
+				 struct gpio_callback *cb, u32_t pins)
```
Please make whitespace fixes in a separate commit

MaureenHelm commented on drivers/sensor/ti_hdc/ti_hdc.c:
```diff
@@ -37,17 +39,22 @@ static int ti_hdc_sample_fetch(struct device *dev, enum sensor_channel chan)
 	buf[0] = TI_HDC_REG_TEMP;
-	if (i2c_write(drv_data->i2c, buf, 1,
-		      DT_TI_HDC_0_BASE_ADDRESS) < 0) {
+	if (i2c_write(drv_data->i2c, buf, 1, DT_TI_HDC_0_BASE_ADDRESS) < 0) {
```
This change seems unrelated to the commit message. Please make whitespace and formatting changes in a separate commit.

MaureenHelm commented on samples/sensor/ti_hdc/CMakeLists.txt:
```diff
@@ -0,0 +1,7 @@
+cmake_minimum_required(VERSION 3.13.1)
```
Please add a license/copyright header

erwango commented on samples/sensor/ti_hdc/sample.yaml:
```diff
@@ -0,0 +1,16 @@
+tests:
+  test_build:
+    build_only: true
+    platform_whitelist: nucleo_l496zg
```
I wonder about the filter here. I think this will (should) compile with any board supporting I2C.
So maybe a `depends on: i2c` would be better here

jfischer-phytec-iot replied:
Indeed, this sample should not be bound to Nucleo STM32L496ZG.

erwango commented on samples/sensor/ti_hdc/sample.yaml:
```diff
@@ -0,0 +1,16 @@
+    build_only: true
+    platform_whitelist: nucleo_l496zg
+    tags: build
+  test_target:
```
I'm not an expert on this part but is this section valid?
Are there rules to add a test_target / harness ? @MaureenHelm any idea?

nikooiko replied:
I use such sample configuration sections for target testing software via a CI and target test-boards. Of course if they don't fit the project rules then I can remove them.

nikooiko replied:
It is removed to simplify this merge request.

nikooiko commented:
I finished all requested changes. @jfischer-phytec-iot @MaureenHelm @erwango can you please recheck to see if I missed anything?

nashif commented:
@nikooiko 

> sensors: renamed hdc1008 driver into ti_hdc  …
> HDC1008 driver is renamed into ti_hdc

You need to explain why you are renaming the driver. there must be a good reason why you are renaming this, so please put this in the commit message.

nikooiko commented:
I believe the rename-driver commit now explains its purpose.

nikooiko commented:
Rebased to master

jfischer-phytec-iot commented on samples/sensor/ti_hdc/sample.yaml:
```diff
@@ -0,0 +1,16 @@
+tests:
+  test_build:
+    build_only: true
+    depends on: i2c
```
s/depends on/depends_on

nikooiko replied:
fixed

jfischer-no committed:
tests: build_all: add TI HDC sensor

Add TI HDC sensor to build_all.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-no committed:
sample: ti_hdc: whitelist reel_board and nucleo_l496zg

Whitelist reel_board and nucleo_l496zg boards.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

MaureenHelm commented on drivers/sensor/ti_hdc/ti_hdc.c:
```diff
@@ -123,12 +131,14 @@ static int ti_hdc_init(struct device *dev)
+	tmp = read16(drv_data->i2c, DT_TI_HDC_0_BASE_ADDRESS,
+		     TI_HDC_REG_DEVICEID);
+	if (tmp != TI_HDC1000_DEVID && tmp != TI_HDC1050_DEVID) {
+		LOG_ERR("Unsupported device ID");
```
Sorry for the nitpick, but the changes here and in `ti_hdc_channel_get` are unrelated to the commit message `sensors: ti_hdc driver wait conversion support` and should be moved to a separate commit

nikooiko replied:
done

nikooiko committed:
sensors: ti_hdc driver wait conversion support

TI_HDC Driver now also supports waiting for conversion to finish instead
of waiting for GPIO interrupt.

Signed-off-by: Nikos Oikonomou <nikoikonomou92@gmail.com>

nikooiko committed:
sensors: ti_hdc now supports 1050 versions

TI_HDC Driver now also supports the 1050 (e.g. hdc1050, hdc1080) temp/rh
sensor versions.

Signed-off-by: Nikos Oikonomou <nikoikonomou92@gmail.com>

nikooiko committed:
sensors: ti_hdc simplified ti_hdc_channel_get

Removed unneeded second temporary variable, used for humidity conversion

Signed-off-by: Nikos Oikonomou <nikoikonomou92@gmail.com>

nikooiko commented:
All changes are done.

MaureenHelm reviewed (approved):
@erwango please revisit

erwango commented on samples/sensor/ti_hdc/nucleo_l496zg.overlay:
```diff
@@ -0,0 +1,15 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+&i2c1 {
```
Is that arduino_spi ?
If it is, maybe this can be moved to something more generic

nikooiko replied:
renamed to **arduino_i2c**, is that ok?

jfischer-phytec-iot replied:
@erwango ping
